File: nixos/common-apps/yggdrasil/update.js
// Run on https://publicpeers.neilalexander.dev/
(function () {
  let elems = document.getElementsByTagName('tr');
  let result = {};
  let country = null;
  for (let i = 0; i < elems.length; i++) {
    if (
      elems[i].className == 'statusgood' ||
      elems[i].className == 'statusavg'
    ) {
      let innerElems = elems[i].getElementsByTagName('td');
      let address = null;
      for (let j = 0; j < innerElems.length; j++) {
        if (innerElems[j].id == 'address') {
          let thisAddress = innerElems[j].innerText;
          if (!thisAddress.startsWith('tls://')) continue;
          address = thisAddress;
          break;
        }
      }
      if (address == null) continue;
      if (country != null) {
        result[country].push(address);
        result[country].sort();
      }
    } else if (elems[i].className == '') {
      let innerElems = elems[i].getElementsByTagName('th');
      for (let j = 0; j < innerElems.length; j++) {
        if (innerElems[j].id == 'country') {
          country = innerElems[j].innerText;
          result[country] = [];
          break;
        }
      }
    }
  }
  document.body.innerHTML =
    '<pre>' + JSON.stringify(result, null, 2) + '</pre>';
})();

File: nixos/common-apps/yggdrasil/update.py
import json
import os
import sys
from typing import Dict, List

import requests
from bs4 import BeautifulSoup

SCRIPT_PATH = os.path.dirname(os.path.realpath(sys.argv[0]))

r = requests.get("https://publicpeers.neilalexander.dev/")
soup = BeautifulSoup(r.text, features="html.parser")

servers: Dict[str, List[str]] = {}
current_country = ""

for row in soup.table.children:
    if row.name == "thead":
        country_elem = row.find(id="country")
        current_country = country_elem.text
    elif row.name == "tbody":
        for server_row in row.find_all(class_="statusgood"):
            address: str = server_row.find(id="address").text
            if not address.startswith("tls://"):
                continue
            if current_country not in servers:
                servers[current_country] = []
            servers[current_country].append(address)
            servers[current_country].sort()

with open(os.path.join(SCRIPT_PATH, "public-peers.json"), "w") as f:
    json.dump(servers, f, indent=2, sort_keys=True)
    f.write("\n")
